### PR TITLE
perf(rfc64): coalesce selected SWM catalog reconciliation

### PR DESCRIPTION
**Pull request by branarakic**
## User impact

Publishing several SWM assets into one selected public Context Graph no longer queues one complete catalog rebuild per asset. The node performs one active author-scope reconcile and, if more inventory mutations arrive during it, one follow-up against the latest durable inventory.

This reduces redundant full-inventory reads and catalog successor work while preserving exact final convergence. It does not subscribe an edge node to more graphs, alter VM inventory, or activate private/unselected graphs.

## Before

```mermaid
sequenceDiagram
    participant A as SWM asset A
    participant B as SWM asset B
    participant C as SWM asset C
    participant Catalog as Exact-set reconciler

    A->>Catalog: Rebuild complete author catalog
    B->>Catalog: Queue another complete rebuild
    C->>Catalog: Queue another complete rebuild
    Note over Catalog: Every asset schedules a full pass
```

## After

```mermaid
sequenceDiagram
    participant A as SWM asset A
    participant B as SWM asset B
    participant C as SWM asset C
    participant Worker as Author-scope worker
    participant Catalog as Exact-set reconciler

    A->>Worker: Mark scope dirty
    Worker->>Catalog: Reconcile durable inventory
    B->>Worker: Mark scope dirty during pass
    C->>Worker: Keep scope dirty
    Worker->>Catalog: One latest-state follow-up
    Catalog-->>Worker: Exact final head
```

## What changed

- Durable inventory mutations return a typed canonical graph, author, and scope key derived inside the validated mutation boundary.
- Promotion and confirmation observers consume that key without rebuilding it from untyped lifecycle input.
- Concurrent requests for one author scope share one completion promise.
- A mutation arriving during a pass causes one bounded latest-state follow-up.
- A failure with no newer mutation exits to the restart/periodic repair lane instead of spinning.
- Different author scopes remain independently concurrent.

## Validation

- Rebased stack: 60/60 targeted RFC-64 tests passed.
- Native production lifecycle: 52 passed.
- Scope coalescing: 7 passed, including burst coalescing, mutation-during-failure retry, and independent-scope concurrency.
- Shared cancellation-safe serialization: 1 passed.
- Agent build, TypeScript checks, package-root checks, lint, and diff checks: passed.

## Stack

Targets #2409 and contains only the coalescing/canonical-key increment.

### CODE REVIEW

File: packages/agent/src/dkg-agent-rfc64-catalog-bootstrap.ts [hunk 1]
```diff
@@ -322,6 +369,16 @@ export class Rfc64CatalogBootstrapMethods extends DKGAgentBase {
...
     state.abortController = abortController;
     try {
+      // Repair locally owned durable author state before any remote discovery
+      // can spend this pass's transport budget.
+      await mapWithConcurrency(
+        state.authorRepairs,
+        MAX_CONCURRENT_TARGETS_V1,
+        async (repair) => {
+          if (state.closed || abortController.signal.aborted) return;
+          await this.reconcileRfc64LocalPublicCatalogAuthorV1(repair);
```
Comment (otReviewAgent) [inline packages/agent/src/dkg-agent-rfc64-catalog-bootstrap.ts]: 🔴 Bug: In-progress local repairs ignore bootstrap cancellation

**What's wrong**
The new repair work is part of the bootstrap run that shutdown drains, but it does not consume the bootstrap abort signal. The pre-call check only prevents new repairs; it cannot stop one already awaiting reconciliation. This defeats the existing close contract and can hang agent shutdown indefinitely.

**Example**
If catalog signing, asset resolution, or reconciliation never settles, calling `closeRfc64PublicCatalogBootstrapV1()` aborts the controller but remains blocked on `state.run`; consequently `DKGAgent.stop()` never reaches catalog-service and node teardown.

**Suggested direction**
Pass cancellation through the repair path and ensure potentially blocking signing, reads, and announcements are bounded or abortable. Shutdown should still drain any durable transition that has crossed its commit boundary.

**For Agents**
Trace cancellation from `runRfc64PublicCatalogBootstrapPassV1` through the local repair and inventory reconciliation paths. Make external work bounded or cooperatively abortable at safe boundaries while preserving already-committed catalog transitions. Add a shutdown test where a repair is in flight and prove close completes without allowing writes after dependent services are closed.

File: packages/agent/test/rfc64-dkg-agent-native-wiring.integration.test.ts [hunk 2]
```diff
@@ -597,21 +597,111 @@ ordinaryNativeWiringDescribe('RFC-64 DKGAgent production native catalog wiring',
     expect(agent.readRfc64PublicCatalogBootstrapStatusV1()).toMatchObject({ pass: 1 });
   });
 
+  it('retries a failed local-author catalog repair without widening graph or author scope', async () => {
```
Comment (otReviewAgent) [inline packages/agent/test/rfc64-dkg-agent-native-wiring.integration.test.ts:693]: 🔴 Bug: The scope-guard test has no out-of-scope candidates

**What's wrong**
Despite claiming that repair does not widen graph or author scope, the test never presents a remote-author target or another graph. It therefore gives false confidence about the new lane's central scope boundary.

**Example**
Add a bootstrap target whose author is absent from listLocalAgents and an ineligible graph target, then assert reconciliation is called only for the selected public graph/local-author tuple and authorRepairs contains only that tuple.

**Suggested direction**
Exercise each security-relevant exclusion with mixed eligible and ineligible targets instead of relying on a single valid tuple.

**For Agents**
In dkg-agent-rfc64-catalog-bootstrap.ts, preserve the intersection of eligible public catalog graphs, catalog-mode targets, and locally registered authors. Extend the native-wiring test with genuine remote-author and ineligible-graph targets, and assert the reconciliation spy's exact calls and resulting authorRepairs entries.

File: packages/agent/src/dkg-agent-rfc64-catalog-bootstrap.ts [hunk 3]
```diff
@@ -378,6 +435,97 @@ export class Rfc64CatalogBootstrapMethods extends DKGAgentBase {
...
+   * Intersect the bounded accepted catalog manifest with authors whose
+   * identities are registered on this node. Remote authors and unselected
+   * CGs can never enter this repair lane.
+   */
+  private planRfc64LocalPublicCatalogAuthorRepairsV1(
+    this: DKGAgent,
+    policies: readonly Rfc64CatalogBootstrapPolicyV1[],
+    targets: readonly Rfc64CatalogBootstrapTargetPlanV1[],
+  ): MutableAuthorRepairStatusV1[] {
+    const autoPublish = this.config.rfc64PublicCatalogAutoPublishPolicy;
```
Comment (otReviewAgent) [inline packages/agent/src/dkg-agent-rfc64-catalog-bootstrap.ts]: 🟡 Issue: Do not duplicate lane eligibility to disambiguate a nullable result

**What's wrong**
The repair lane only knows that `null` means “no inventory” because it reimplements the canonical auto-publish admission rules first. This creates two sources of truth for rollout and policy eligibility and leaks auto-publish internals into bootstrap orchestration. The abstractions can drift whenever lane admission changes.

**Example**
A future eligibility rule added to `resolveRfc64AcceptedPublicRootLaneV1` must also be copied into this planner. If it is not, the bootstrap repair status can call an ineligible lane and report its ambiguous `null` result as `no-inventory`.

**Suggested direction**
Keep activation, accepted-policy, and inventory interpretation in the auto-publish layer. Expose either a canonical repair-candidate resolver or a dedicated restart-repair operation with a discriminated result such as inactive/no-inventory/reconciled. Bootstrap should only schedule and retry the resulting local repair tasks, allowing these copied filters and casts to disappear.

**For Agents**
Refactor the restart-repair boundary across `dkg-agent-rfc64-catalog-bootstrap.ts` and `dkg-agent-rfc64-catalog-auto-publish.ts`. Preserve repair of only operator-pinned, locally controlled, selected public catalog authors. Add a canonical typed repair/candidate API that owns lane eligibility and distinguishes inactive from missing-inventory and successful outcomes; prove selected local repair still runs while private, unselected, and remote-author targets remain excluded.

File: packages/agent/test/rfc64-dkg-agent-native-wiring.integration.test.ts [hunk 4]
```diff
@@ -859,6 +970,49 @@ ordinaryNativeWiringDescribe('RFC-64 DKGAgent production native catalog wiring',
...
+    expect(restarted.rfc64SwmAuthorInventoryShadowStatusV1()).toMatchObject({
+      attemptedRemovals: 1,
+      appliedRemovals: 1,
+      absentRemovals: 0,
+      failed: 0,
+    });
+    reconcileSpy.mockRestore();
+    await restarted.stop();
+    agents.splice(agents.indexOf(restarted), 1);
+    restarted = await startNativeAgentWithOptions({
```
Comment (otReviewAgent) [inline packages/agent/test/rfc64-dkg-agent-native-wiring.integration.test.ts]: 🟡 Issue: Split the second restart-repair lifecycle out of the 550-line test

**What's wrong**
This change further expands an already oversized 4.5k-line integration file and embeds another complete restart lifecycle inside a single roughly 550-line test. Fixture state and failure diagnosis now depend on many unrelated earlier phases, making the new behavior expensive to understand or modify.

**Example**
The same `it` now covers restricted-share filtering, inventory creation, workspace drift, first-restart recovery, idempotent promotion, failed post-removal reconciliation, second-restart recovery, and later concurrency behavior while mutating one `restarted` variable across phases.

**Suggested direction**
Move restart-repair coverage into focused scenarios, preferably in a dedicated integration file, and extract the repeated repair-enabled agent startup fixture. Each test should own one failure/restart/recovery lifecycle rather than appending more phases to this monolith.

**For Agents**
Decompose `rfc64-dkg-agent-native-wiring.integration.test.ts` around the changed restart-repair scenario. Extract durable-inventory seeding and repair-enabled restart helpers, then make the first-restart and post-removal second-restart recovery paths focused tests. Preserve the existing assertions for catalog versions, row counts, and inventory mutation statistics.

File: packages/agent/src/dkg-agent-rfc64-catalog-bootstrap.ts [hunk 5]
```diff
@@ -378,6 +435,97 @@ export class Rfc64CatalogBootstrapMethods extends DKGAgentBase {
...
+  ): Promise<void> {
+    repair.attempts += 1;
+    try {
+      const reconciled = await this.reconcileRfc64PublicCatalogFromSwmInventoryV1({
+        contextGraphId: repair.contextGraphId,
+        authorAddress: repair.authorAddress,
+      });
+      if (reconciled === null) {
+        Object.assign(repair, {
+          outcome: 'no-inventory',
```
Comment (otReviewAgent) [inline packages/agent/src/dkg-agent-rfc64-catalog-bootstrap.ts]: 🟡 Issue: The new no-inventory status contract is untested

**What's wrong**
The PR exports no-inventory as an observable repair outcome, but no test executes that branch. A regression that reports pending, failed, or reconciled for a node with no durable inventory would pass the suite.

**Example**
Have reconcileRfc64PublicCatalogFromSwmInventoryV1 resolve null on the first pass and assert outcome is no-inventory, attempts is 1, and the digest/version/row-count/error fields are null. A separate empty signed inventory can verify it remains reconciled with row count 0 rather than no-inventory.

**Suggested direction**
Add a focused status test for a null reconciliation result, ideally alongside the distinct empty-inventory success case.

**For Agents**
Add coverage near the new local-author repair tests for the null reconciliation branch in dkg-agent-rfc64-catalog-bootstrap.ts. Prove the public status distinguishes a missing inventory from an existing empty inventory.

File: packages/agent/src/rfc64/swm-inventory-shadow-runtime-v1.ts [hunk 6]
```diff
@@ -73,6 +111,38 @@ export class Rfc64SwmInventoryShadowRuntimeV1 {
...
+  /**
+   * Coalesce author-scope convergence onto the latest durable inventory. A
+   * mutation arriving during a pass marks the scope dirty and causes exactly
+   * one follow-up pass; callers share completion of the latest requested set.
+   */
+  requestScopeReconciliation(
+    scopeKey: string,
+    reconcile: () => Promise<void>,
+  ): Promise<void> {
+    const existing = this.#scopeReconciliations.get(scopeKey);
```
Comment (otReviewAgent) [inline packages/agent/src/rfc64/swm-inventory-shadow-runtime-v1.ts]: 🔴 Bug: A reconciliation request can be lost while the previous worker is settling

**What's wrong**
The state remains discoverable in the map until `.finally()` runs, even though `runScopeReconciliation` may already have exited. A new durable mutation in that interval sets `dirty`, but no worker remains to observe it. Its caller also receives the old completion as if the latest inventory had been reconciled, potentially leaving the public catalog stale until another mutation or restart repairs it.

**Example**
Start a reconciliation whose callback returns a gated promise. Resolve the gate, then enqueue another `requestScopeReconciliation` call in a microtask before the first completion finalizer runs. The second call returns the first promise, but the callback runs only once; the newly requested catalog pass is lost.

**Suggested direction**
Make the clean-state decision and removal from `#scopeReconciliations` atomic from the event loop's perspective, or track a closed state and start a replacement worker instead of returning a completion that can no longer process `dirty`.

**For Agents**
Update `Rfc64SwmInventoryShadowRuntimeV1` so a worker is removed or marked closed synchronously when it observes a clean state, before its promise settles. A request seeing a closed worker must create a new worker. Preserve coalescing and bounded retry behavior. Add a test that requests reconciliation at both successful and rejected promise-settlement boundaries and proves the latest request executes.

File: packages/agent/test/encrypt-inline-policy.test.ts [hunk 7]
```diff
@@ -999,6 +1000,11 @@ describe('DKGAgent.publishQueuedKnowledgeAssetVmPublish inline encryption routin
       attempts: 1,
       headObjectDigest: null,
       error: null,
+      reconciliation: {
```
Comment (otReviewAgent) [inline packages/agent/test/encrypt-inline-policy.test.ts]: 🔴 Bug: Queued-publish test silently exercises a failed reconciliation

**What's wrong**
The newly added non-null reconciliation fixture enters the changed behavior but the harness lacks the method that behavior calls. Its TypeError is deliberately swallowed by the production failure boundary, so this test gives false confidence about the queued-publish reconciliation wiring.

**Example**
Add `agentLike.reconcileRfc64PublicCatalogFromSwmInventoryV1 = recorder(async () => null)`, then assert it receives `{ contextGraphId: 'public-cg', authorAddress: QUEUED_TEST_AUTHOR }`. The current test would instead record a swallowed TypeError.

**Suggested direction**
Stub and assert the canonical reconciliation method, and optionally verify that no lifecycle warning was emitted, so the test cannot pass through the observer's failure boundary.

**For Agents**
Update the confirmed queued-publish test and its harness in `encrypt-inline-policy.test.ts`. Preserve the assertions that deprecated catalog-authoring methods are unused, but provide the canonical reconciliation method and prove it is called successfully with the request metadata.

File: packages/agent/src/rfc64/swm-inventory-shadow-runtime-v1.ts [hunk 8]
```diff
@@ -73,6 +111,38 @@ export class Rfc64SwmInventoryShadowRuntimeV1 {
     }
   }
 
+  /**
+   * Coalesce author-scope convergence onto the latest durable inventory. A
+   * mutation arriving during a pass marks the scope dirty and causes exactly
+   * one follow-up pass; callers share completion of the latest requested set.
+   */
+  requestScopeReconciliation(
```
Comment (otReviewAgent) [inline packages/agent/src/rfc64/swm-inventory-shadow-runtime-v1.ts]: 🟡 Issue: Make reconciliation a scope-owned coordinator instead of passing replaceable commands

**What's wrong**
The new abstraction splits one reconciliation identity across a raw key, captured callback values, and an exported mutation-result payload. Nothing guarantees those pieces describe the same scope, yet the coalescer replaces callbacks solely by string-key equality. This leaks orchestration into a public result model and makes the implementation depend on an undocumented invariant rather than one canonical owner.

**Example**
A future caller can supply a request for context graph A with a scope key derived from graph B. The runtime will coalesce it under B while the callback reconciles A. Likewise, two different callbacks submitted under one key cause the later callback to silently replace the earlier one.

**Suggested direction**
Give the coordinator a single reconciliation function and expose `request(scope)` using a typed canonical scope whose key is derived internally. Keep `Rfc64SwmAuthorInventoryShadowMutationResultV1` focused on mutation status/telemetry, and have one internal observer helper request reconciliation for eligible outcomes. This would remove the mutable callback, raw-key/closure alignment invariant, duplicated observer blocks, overload set, cast, and unrelated mock churn.

**For Agents**
Refactor the RFC-64 shadow runtime and auto-publish observer paths. Preserve the current status values, retry/coalescing semantics, and one trailing reconciliation pass. Introduce a canonical typed scope identity/key builder, make a dedicated coordinator own the reconciliation operation, and keep scheduling metadata out of the exported mutation result. The runtime tests should prove same-scope coalescing and failure-triggered retry without passing replaceable callbacks; existing non-catalog mutation-result mocks should no longer need reconciliation fields.

File: packages/agent/src/dkg-agent-rfc64-catalog-auto-publish.ts [hunk 9]
```diff
@@ -121,14 +122,46 @@ export interface RecordRfc64PublicCatalogAssetParamsV1 {
...
 ): Rfc64SwmAuthorInventoryShadowMutationResultV1 {
-  return Object.freeze({ status, action, attempts, headObjectDigest, error });
+  return Object.freeze({
+    status,
+    action,
+    attempts,
+    headObjectDigest,
+    error,
+    reconciliation,
+  }) as Rfc64SwmAuthorInventoryShadowMutationResultV1;
```
Comment (otReviewAgent) [inline packages/agent/src/dkg-agent-rfc64-catalog-auto-publish.ts]: 🟡 Issue: Make result variants precise enough to eliminate the assertion-based factory

**What's wrong**
The new discriminated union only correlates status with action and reconciliation; it leaves the error invariant loose. The factory then adds three positional overloads and asserts the assembled object into the result type. This creates significant type machinery without establishing a trustworthy boundary, and future field or variant changes can be hidden by the cast.

**Example**
TypeScript accepts an exported result shaped like `{ status: 'applied', action: 'upsert', attempts: 1, headObjectDigest: null, error: 'failure', reconciliation: request }`, even though the factory overloads imply that an applied result must have `error: null`.

**Suggested direction**
Model each semantic variant explicitly and construct it through an object-shaped input checked against the union. This should let the compiler verify status, action, error, and reconciliation together without overload duplication or a terminal cast.

**For Agents**
Split dormant, failed, and successful results into truthful variants, including `error: string` for failure and `error: null` otherwise. Replace the positional overload/cast with checked object literals or an object-argument factory. Preserve all existing statuses and action restrictions; type-level checks should reject contradictory combinations.

File: packages/agent/src/rfc64/swm-inventory-shadow-runtime-v1.ts (hunk 8 above)
Comment (branarakic) [reply packages/agent/src/rfc64/swm-inventory-shadow-runtime-v1.ts]: Addressed in 60469e11c: one agent-owned reconciler is injected into the coordinator, callers pass a typed CG and author scope, the coordinator derives its key internally, and scheduling metadata was removed from the mutation result.

File: packages/agent/src/dkg-agent-rfc64-catalog-bootstrap.ts (hunk 1 above)
Comment (branarakic) [reply packages/agent/src/dkg-agent-rfc64-catalog-bootstrap.ts]: Addressed in parent 3ac8a8035: restart repair now accepts and propagates cancellation through admission, persistence reads, signing, queue waits, exact-set publication, and announcement, with an in-flight close test.

File: packages/agent/test/rfc64-dkg-agent-native-wiring.integration.test.ts (hunk 2 above)
Comment (branarakic) [reply packages/agent/test/rfc64-dkg-agent-native-wiring.integration.test.ts:693]: Addressed in parent 3ac8a8035: canonical repair admission now requires the selected RFC-64 root lane and the local eligible catalog author, and explicitly excludes inactive, wrong-policy, restricted, and remote-author scopes.

File: packages/agent/test/rfc64-dkg-agent-native-wiring.integration.test.ts (hunk 4 above)
Comment (branarakic) [reply packages/agent/test/rfc64-dkg-agent-native-wiring.integration.test.ts]: Addressed in parent 3ac8a8035: the monolithic restart case was split into focused addition and removal lifecycle tests while retaining the broader integration coverage.

File: packages/agent/src/dkg-agent-rfc64-catalog-bootstrap.ts (hunk 3 above)
Comment (branarakic) [reply packages/agent/src/dkg-agent-rfc64-catalog-bootstrap.ts]: Addressed in parent 3ac8a8035: restart repair now has typed candidate and result boundaries, with explicit inactive, no-inventory, and reconciled outcomes.

File: packages/agent/src/rfc64/swm-inventory-shadow-runtime-v1.ts (hunk 6 above)
Comment (branarakic) [reply packages/agent/src/rfc64/swm-inventory-shadow-runtime-v1.ts]: Addressed in 60469e11c: a worker is synchronously closed and removed before its promise settles. Requests at either successful or rejected settlement boundaries create a distinct replacement worker; both races are covered by focused regression tests.

File: packages/agent/src/dkg-agent-rfc64-catalog-bootstrap.ts (hunk 5 above)
Comment (branarakic) [reply packages/agent/src/dkg-agent-rfc64-catalog-bootstrap.ts]: Addressed in parent 3ac8a8035: missing inventory is now a distinct typed no-inventory result rather than being conflated with an empty authoritative inventory.

File: packages/agent/test/encrypt-inline-policy.test.ts (hunk 7 above)
Comment (branarakic) [reply packages/agent/test/encrypt-inline-policy.test.ts]: Addressed in 60469e11c: the queued-publish harness now attaches the real reconciliation helper, stubs the canonical reconciliation method, asserts its exact CG and author scope, and verifies that no lifecycle warning was swallowed.

File: packages/agent/test/rfc64-dkg-agent-native-wiring.integration.test.ts (hunk 4 above)
Comment (otReviewAgent) [reply packages/agent/test/rfc64-dkg-agent-native-wiring.integration.test.ts]: 🟡 Issue: Isolate the retry scenario from the long restart narrative

**What's wrong**
This adds another stateful story in the middle of a test that already covers several lifecycle phases. The new story changes all subsequent state, coupling otherwise separate behaviors and making failures or future edits harder to localize.

**Example**
Adding this block cascades into later expectation changes from inventory version `0` to `2`, catalog version `1` to `3`, and final catalog version `2` to `4`.

**Suggested direction**
Give the failure-and-retry behavior a fresh fixture or focused helper-backed test. That keeps catalog versions and mutation counters local to the behavior they describe and prevents future scenarios from causing cascading expectation rewrites.

**For Agents**
Extract reusable SWM inventory seeding/setup from this describe block and move the simulated reconciliation-failure/idempotent-removal retry into its own `it`. Keep the existing restricted-share/restart/removal scenario on its original state progression, and run both integration cases.

File: packages/agent/src/rfc64/swm-inventory-shadow-runtime-v1.ts (hunk 6 above)
Comment (otReviewAgent) [reply packages/agent/src/rfc64/swm-inventory-shadow-runtime-v1.ts]: 🟡 Issue: No regression test covers a request arriving during the follow-up pass

**What's wrong**
The new coalescer promises convergence to the latest requested inventory, but its tests only cover requests arriving during the initial pass. If the loop later regressed to allow at most one follow-up, a mutation arriving during that follow-up could leave the catalog stale while every existing test remained green.

**Example**
Gate reconciliation passes 1 and 2 separately. Request B while pass 1 is blocked, release pass 1, request C while pass 2 is blocked, then release pass 2. The test should require three passes and verify all callers share the same completion promise. An implementation capped at one follow-up would pass the current tests but drop C.

**Suggested direction**
Add a focused runtime test that blocks both the initial and follow-up reconciliation passes and submits another same-scope request during pass two.

**For Agents**
Extend `packages/agent/test/rfc64-swm-inventory-shadow-runtime-v1.test.ts` with a two-gate test around `requestScopeReconciliation`. Prove that a mutation arriving during the follow-up pass marks the scope dirty again, triggers a third pass, and remains covered by the shared completion promise.

File: packages/agent/src/rfc64/swm-inventory-shadow-runtime-v1.ts [hunk 10]
```diff
@@ -44,10 +67,22 @@ export class Rfc64SwmInventoryShadowRuntimeV1 {
   readonly #inFlight = new Set<Promise<void>>();
   readonly #assetTails = new Map<string, Promise<void>>();
   readonly #scopeTails = new Map<string, Promise<void>>();
+  readonly #scopeReconciliations = new Map<string, ScopeReconciliationStateV1>();
```
Comment (otReviewAgent) [inline packages/agent/src/rfc64/swm-inventory-shadow-runtime-v1.ts]: 🟡 Issue: Split catalog reconciliation out of the shadow-runtime grab bag

**What's wrong**
The class is becoming a general-purpose bag of unrelated process-local mechanisms. Catalog convergence now adds its own callback, map, mutable state machine, and lifecycle methods to a runtime already responsible for several independent concerns. This increases coupling and makes independent use or testing require meaningless dependencies.

**Example**
Every pre-existing `new Rfc64SwmInventoryShadowRuntimeV1()` call becomes `new Rfc64SwmInventoryShadowRuntimeV1(async () => undefined)` even though those tests never reconcile a catalog.

**Suggested direction**
Introduce a focused keyed reconciliation/coalescing component and keep the shadow runtime dependency-free. Inside that component, map membership can represent the active state and `try/finally` can own cleanup, eliminating the unobservable `closed` mode and duplicated close paths.

**For Agents**
Extract the keyed latest-state reconciliation runner from `swm-inventory-shadow-runtime-v1.ts` and compose it beside the shadow runtime in the agent factory. Preserve same-scope promise sharing, one dirty follow-up, unrelated-scope concurrency, and retry-after-settlement behavior in the existing runtime tests.

Comment (branarakic) [reply packages/agent/src/rfc64/swm-inventory-shadow-runtime-v1.ts]: Addressed in 6a5b58bd2. Catalog reconciliation now lives in a focused `Rfc64SwmCatalogReconciliationCoordinatorV1` composed beside the dependency-free shadow runtime. The coordinator owns keyed active-scope state, coalescing, bounded retry, and settlement cleanup via map membership/`finally`; its four reconciliation tests moved to a dedicated suite. The shadow runtime is back to observer scheduling, asset ordering, scope serialization, VM tombstones, and metrics only.

File: packages/agent/src/dkg-agent-rfc64-catalog-auto-publish.ts (hunk 9 above)
Comment (branarakic) [reply packages/agent/src/dkg-agent-rfc64-catalog-auto-publish.ts]: Resolved by the rebased stack. The assertion-based overloaded result factory that this thread referenced is no longer part of #2410's diff; this increment now only carries canonical scope-key propagation and the focused coalescing coordinator. The current 6a5b58bd2 head passes the 60-test RFC-64 stack and type/package checks.

File: packages/agent/src/rfc64/swm-catalog-reconciliation-coordinator-v1.ts [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+   * one latest-state follow-up. Callers for one active scope share completion.
+   */
+  request(
+    requestedScope: Readonly<Rfc64SwmCatalogReconciliationScopeV1>,
+  ): Promise<void> {
+    const scope = Object.freeze({
+      contextGraphId: requestedScope.contextGraphId,
+      authorAddress: requestedScope.authorAddress.toLowerCase() as EvmAddressV1,
+    });
+    const scopeKey = rfc64SwmCatalogReconciliationScopeKeyV1(scope);
```
Comment (otReviewAgent) [inline packages/agent/src/rfc64/swm-catalog-reconciliation-coordinator-v1.ts:49]: 🟡 Issue: Per-scope isolation is not tested

**What's wrong**
The central isolation property of the new coalescer has no regression coverage. A keying regression could merge unrelated authors or context graphs, leaving one catalog unreconciled while all current tests remain green.

**Example**
Block reconciliation for scope A, then request scope B. Scope B should start independently and receive a different completion promise. A broken key that omitted the context graph or author would still pass the current suite.

**Suggested direction**
Add an overlapping multi-scope test covering both context-graph and author separation.

**For Agents**
Add a coordinator test using two distinct scopes with the first reconciliation gated. Prove the second scope enters before the first is released, receives a different promise, and both exact scopes are reconciled.

File: packages/agent/src/rfc64/swm-catalog-reconciliation-coordinator-v1.ts [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+    this.#activeScopes.set(scopeKey, state);
+    return completion;
+  }
+
+  private async run(
+    scopeKey: string,
+    state: ScopeReconciliationStateV1,
+  ): Promise<void> {
+    try {
+      while (state.dirty) {
```
Comment (otReviewAgent) [inline packages/agent/src/rfc64/swm-catalog-reconciliation-coordinator-v1.ts:74]: 🟡 Issue: Mutations arriving during the follow-up pass are unverified

**What's wrong**
The tests verify requests arriving during the initial pass, but not during the follow-up. Losing a mutation in that later window would leave the signed catalog behind durable inventory without failing the current suite.

**Example**
Gate pass 1, request version 2, release pass 1, gate pass 2, then request version 3. The completion should run a third pass and observe versions `[1, 2, 3]`; an implementation limited to one trailing pass would incorrectly stop at version 2 but pass the current tests.

**Suggested direction**
Add a three-pass concurrency regression test proving that requests during any active pass remain dirty and are not lost.

**For Agents**
Extend the coordinator unit suite with a gated second reconciliation. Submit another same-scope request during that follow-up and assert all callers share completion and a third latest-state pass runs.

File: packages/agent/src/dkg-agent-rfc64-catalog-auto-publish.ts [hunk 13]
```diff
@@ -322,18 +400,92 @@ export class Rfc64CatalogAutoPublishMethods extends DKGAgentBase {
...
     if (lane === null) return null;
+    return this.reconcileRfc64PublicCatalogFromSwmInventoryLaneV1(lane, params);
+  }
+
+  /**
+   * Canonical restart-repair admission. Bootstrap supplies only a bounded
+   * manifest tuple; this owner applies rollout, policy, selection, root-lane,
+   * and local-author rules as one capability boundary.
+   */
+  resolveRfc64LocalPublicCatalogAuthorRepairCandidateV1(
```
Comment (otReviewAgent) [inline packages/agent/src/dkg-agent-rfc64-catalog-auto-publish.ts:424]: 🟡 Issue: Decompose this class before pushing it past 1,000 lines

**What's wrong**
This PR crosses the explicit 1,000-line threshold and makes an already broad class responsible for another lifecycle. The resulting module no longer has a readily scannable ownership boundary.

**Example**
A maintainer changing restart repair now has to navigate the same 1,099-line class that also owns SWM inventory mutation, lifecycle observation, catalog reconciliation, asset resolution, signing, and rollout-lane selection.

**Suggested direction**
Give local-author restart repair, lane admission, and reconciliation orchestration a focused owner instead of continuing to expand the auto-publish class.

**For Agents**
Decompose packages/agent/src/dkg-agent-rfc64-catalog-auto-publish.ts before merging. Extract the new local-author repair capability and its lane-decision/result types into a focused RFC-64 module or mixin while preserving the existing public agent methods. Existing reconciliation and restart tests should continue proving identical outcomes.

File: packages/agent/src/dkg-agent-rfc64-catalog-bootstrap.ts [hunk 14]
```diff
@@ -322,6 +370,19 @@ export class Rfc64CatalogBootstrapMethods extends DKGAgentBase {
     const abortController = new AbortController();
     state.abortController = abortController;
     try {
+      // Repair locally owned durable author state before any remote discovery
+      // can spend this pass's transport budget.
+      await mapWithConcurrency(
```
Comment (otReviewAgent) [inline packages/agent/src/dkg-agent-rfc64-catalog-bootstrap.ts:375]: 🟡 Issue: Do not make local repair a blocking prelude to remote bootstrap

**What's wrong**
The new local repair phase is serialized ahead of unrelated provider discovery and target synchronization. This couples independent workflows and makes the bootstrap pass harder to evolve or reason about.

**Example**
One slow local catalog repair delays every configured provider dial and every remote bootstrap target, even when those targets belong to unrelated context graphs and authors.

**Suggested direction**
Run the independent local-repair and remote-bootstrap lanes concurrently, or schedule them through one bounded prioritized queue if a shared global budget is required.

**For Agents**
Restructure runRfc64PublicCatalogBootstrapPassV1 so local repair and remote bootstrap are sibling tasks. Preserve bounded concurrency and shared cancellation; use parallel lanes or one bounded scheduler with explicit priorities. Add an orchestration test showing remote work can begin while local repair is pending.

File: devnet/rfc64-gate1-public-open/adapter-process.ts [hunk 15]
```diff
@@ -647,6 +679,14 @@ function requiredString(value: unknown, label: string): string {
   return value;
 }
 
+function assertSafeIri(value: string, label: string): string {
```
Comment (otReviewAgent) [inline devnet/rfc64-gate1-public-open/adapter-process.ts:682]: 🟡 Issue: Reuse the canonical SPARQL IRI validator

**What's wrong**
The change duplicates an existing canonical helper in two new locations, creating three sources of truth for the same SPARQL boundary.

**Example**
A future update to the canonical unsafe-IRI character policy would require discovering and updating both new devnet copies or the rollout certificate would silently use different validation rules.

**Suggested direction**
Delete both bespoke validators and route validation through the core helper that already owns this invariant.

**For Agents**
Import the canonical assertSafeIri from @origintrail-official/dkg-core in adapter-process.ts and rollout-store-fixture.ts. Preserve boundary-specific labels by wrapping/rethrowing only if the custom diagnostic is necessary; remove both local regex implementations.

File: packages/agent/src/rfc64/public-catalog-service-v1.ts [hunk 16]
```diff
@@ -874,12 +875,18 @@ export class Rfc64PublicCatalogServiceV1 {
...
     const announcedPeers: string[] = [];
     const failedPeers: Array<{ peerId: string; error: string }> = [];
     for (const peerId of peers) {
+      if (signal?.aborted) break;
       try {
-        await this.#transport.announceCatalogHead(peerId, announcement, this.#sendOptions());
+        await this.#transport.announceCatalogHead(
+          peerId,
+          announcement,
+          this.#sendOptions(signal),
```
Comment (otReviewAgent) [inline packages/agent/src/rfc64/public-catalog-service-v1.ts:888]: 🟡 Issue: In-flight announcement cancellation is not covered

**What's wrong**
The change adds cancellation to catalog fan-out, but no test passes an AbortSignal through this service path. A regression that drops the signal from the transport options—or continues fan-out after cancellation—would leave the suite green and could delay bootstrap shutdown until the transport timeout.

**Example**
Use a transport whose first `announceCatalogHead` waits for `options.signal` to abort, call the service with two peers, abort the controller, and assert the call settles promptly, the signal reached the first transport call, and the second peer was not attempted.

**Suggested direction**
Add a regression test using a signal-aware hanging transport rather than relying only on the earlier-stage bootstrap cancellation test.

**For Agents**
Add a focused test in `packages/agent/test/rfc64-public-catalog-service-v1.test.ts` around `announceCatalogHead`. Exercise an in-flight transport request with an AbortSignal and prove prompt settlement plus suppression of later peer attempts.

**Comment by branarakic** [comment]
Current-stack audit against #2409 head 1361e1550: this PR’s reconciliation coordinator is now subsumed by #2409’s scope-owned projection supervisor. Carrying the coordinator forward would duplicate coalescing and reintroduce an awaited reconciliation inside the detached observer, contrary to #2409’s non-blocking user-operation boundary. The canonical scope isolation, arbitrary-pass dirty-loop behavior, restart retry, remote/local writer serialization, cancellation, and lifecycle drain are now covered on #2409. The only residual here is additional idempotent-removal regression coverage, which does not justify a competing runtime. Recommendation: do not merge the current head; treat #2410 as superseded after #2409 lands. I am leaving the PR open and unchanged pending owner approval to close it.

**Comment by branarakic-agent** [comment]
Closing as superseded. The current stack audit recorded that #2409’s scope-owned projection supervisor now owns this behavior; carrying this coordinator forward would duplicate coalescing and reintroduce the blocking observer path. Any uniquely useful idempotent-removal regression can be recovered in a focused follow-up.
